File: sdks/csharp/src/Event.cs
using System;
using System.Collections.Generic;
using SpacetimeDB.ClientApi;

namespace SpacetimeDB
{
    public interface IEventContext
    {

    }

    public interface IReducerEventContext
    {

    }

    public interface ISubscriptionEventContext
    {

    }

    public interface IErrorContext
    {
        public Exception Event { get; }
    }

    // The following underscores are needed to work around c#'s unified type-and-function
    // namespace.

    /// <summary>
    /// The <c>DbContext</c> trait, which mediates access to a remote module.
    ///
    /// <c>DbContext</c> is implemented by <c>DbConnection</c> and <c>EventContext</c>,
    /// both defined in your module-specific codegen.
    /// </summary>
    public interface IDbContext<DbView, RemoteReducers, SetReducerFlags_, SubscriptionBuilder_>
    {
        /// <summary>
        /// Access to tables in the client cache, which stores a read-only replica of the remote database state.
        ///
        /// The returned <c>DbView</c> will have a method to access each table defined by the module.
        /// </summary>
        public DbView Db { get; }

        /// <summary>
        /// Access to reducers defined by the module.
        ///
        /// The returned <c>RemoteReducers</c> will have a method to invoke each reducer defined by the module,
        /// plus methods for adding and removing callbacks on each of those reducers.
        /// </summary>
        public RemoteReducers Reducers { get; }

        /// <summary>
        /// Access to setters for per-reducer flags.
        ///
        /// The returned <c>SetReducerFlags</c> will have a method to invoke,
        /// for each reducer defined by the module,
        /// which call-flags for the reducer can be set.
        /// </summary>
        public SetReducerFlags_ SetReducerFlags { get; }

        /// <summary>
        /// Returns <c>true</c> if the connection is active, i.e. has not yet disconnected.
        /// </summary>
        public bool IsActive { get; }

        /// <summary>
        /// Close the connection.
        ///
        /// Throws an error if the connection is already closed.
        /// </summary>
        public void Disconnect();

        /// <summary>
        /// Start building a subscription.
        /// </summary>
        /// <returns>A builder-pattern constructor for subscribing to queries,
        /// causing matching rows to be replicated into the client cache.</returns>
        public SubscriptionBuilder_ SubscriptionBuilder();

        /// <summary>
        /// Get the <c>Identity</c> of this connection.
        ///
        /// This method returns null if the connection was constructed anonymously
        /// and we have not yet received our newly-generated <c>Identity</c> from the host.
        /// </summary>
        public Identity? Identity { get; }

        /// <summary>
        /// Get this connection's <c>ConnectionId</c>.
        /// </summary>
        public ConnectionId ConnectionId { get; }
    }

    public interface IReducerArgs : BSATN.IStructuralReadWrite
    {
        string ReducerName { get; }
    }

    [Type]
    public partial record Status : TaggedEnum<(
        Unit Committed,
        string Failed,
        Unit OutOfEnergy
    )>;

    public record ReducerEvent<R>(
        Timestamp Timestamp,
        Status Status,
        Identity CallerIdentity,
        ConnectionId? CallerConnectionId,
        U128? EnergyConsumed,
        R Reducer
    );

    public record Event<R>
    {
        private Event() { }

        public record Reducer(ReducerEvent<R> ReducerEvent) : Event<R>;
        public record SubscribeApplied : Event<R>;
        public record UnsubscribeApplied : Event<R>;
        public record SubscribeError(Exception Exception) : Event<R>;
        public record UnknownTransaction : Event<R>;
    }


    public interface ISubscriptionHandle
    {
        void OnApplied(ISubscriptionEventContext ctx, SubscriptionAppliedType state);
        void OnError(IErrorContext ctx);
        void OnEnded(ISubscriptionEventContext ctx);
    }

    /// <summary>
    /// An applied subscription can either be a new-style subscription (with a query ID),
    /// or a legacy subscription (no query ID).
    /// </summary>
    [Type]
    public partial record SubscriptionAppliedType : TaggedEnum<(
        QueryId Active,
        Unit LegacyActive)>
    { }

    /// <summary>
    /// State flow chart:
    /// <c>
    ///           |
    ///           v
    ///        Pending
    ///        |     |
    ///        v     v
    ///     Active  LegacyActive
    ///        |
    ///        v
    ///     Ended
    /// </c>
    /// </summary>
    [Type]
    public partial record SubscriptionState
        : TaggedEnum<(Unit Pending, QueryId Active, Unit LegacyActive, Unit Ended)>
    { }

    public class SubscriptionHandleBase<SubscriptionEventContext, ErrorContext> : ISubscriptionHandle
        where SubscriptionEventContext : ISubscriptionEventContext
        where ErrorContext : IErrorContext
    {
        private readonly IDbConnection conn;
        private readonly Action<SubscriptionEventContext>? onApplied;
        private readonly Action<ErrorContext, Exception>? onError;
        private Action<SubscriptionEventContext>? onEnded;

        private SubscriptionState state;

        /// <summary>
        /// Whether the subscription has ended.
        /// </summary>
        public bool IsEnded
        {
            get
            {
                return state is SubscriptionState.Ended;
            }
        }

        /// <summary>
        /// Whether the subscription is active.
        /// </summary>
        public bool IsActive
        {
            get
            {
                return state is SubscriptionState.Active || state is SubscriptionState.LegacyActive;
            }
        }

        void ISubscriptionHandle.OnApplied(ISubscriptionEventContext ctx, SubscriptionAppliedType type)
        {
            if (type is SubscriptionAppliedType.Active active)
            {
                state = new SubscriptionState.Active(active.Active_);
            }
            else if (type is SubscriptionAppliedType.LegacyActive)
            {
                state = new SubscriptionState.LegacyActive(new());
            }
            onApplied?.Invoke((SubscriptionEventContext)ctx);
        }

        void ISubscriptionHandle.OnEnded(ISubscriptionEventContext ctx)
        {
            state = new SubscriptionState.Ended(new());
            onEnded?.Invoke((SubscriptionEventContext)ctx);
        }

        void ISubscriptionHandle.OnError(IErrorContext ctx)
        {
            state = new SubscriptionState.Ended(new());
            onError?.Invoke((ErrorContext)ctx, ctx.Event);
        }

        /// <summary>
        /// Construct a legacy subscription handle.
        /// </summary>
        /// <param name="conn"></param>
        /// <param name="onApplied"></param>
        /// <param name="onError"></param>
        /// <param name="querySqls"></param>
        protected SubscriptionHandleBase(IDbConnection conn, Action<SubscriptionEventContext>? onApplied, string[] querySqls)
        {
            state = new SubscriptionState.Pending(new());
            this.conn = conn;
            this.onApplied = onApplied;
            conn.LegacySubscribe(this, querySqls);
        }

        /// <summary>
        /// Construct a subscription handle.
        /// </summary>
        /// <param name="conn"></param>
        /// <param name="onApplied"></param>
        /// <param name="onError"></param>
        /// <param name="querySql"></param>
        protected SubscriptionHandleBase(
            IDbConnection conn,
            Action<SubscriptionEventContext>? onApplied,
            Action<ErrorContext, Exception>? onError,
            string[] querySqls
        )
        {
            state = new SubscriptionState.Pending(new());
            this.onApplied = onApplied;
            this.onError = onError;
            this.conn = conn;
            conn.Subscribe(this, querySqls);
        }

        /// <summary>
        /// Unsubscribe from the query controlled by this subscription handle.
        /// 
        /// Calling this more than once will result in an exception.
        /// </summary>
        public void Unsubscribe()
        {
            UnsubscribeThen(null);
        }

        /// <summary>
        /// Unsubscribe from the query controlled by this subscription handle,
        /// and call onEnded when its rows are removed from the client cache.
        /// </summary>
        public void UnsubscribeThen(Action<SubscriptionEventContext>? onEnded)
        {
            if (state is not SubscriptionState.Active)
            {
                throw new Exception("Cannot unsubscribe from inactive subscription.");
            }
            if (this.onEnded != null)
            {
                throw new Exception("Unsubscribe already called.");
            }
            if (onEnded == null)
            {
                // We need to put something in there to use this as a boolean.
                onEnded = (ctx) => { };
            }
            this.onEnded = onEnded;
        }
    }
}

File: sdks/rust/src/event.rs
//! The [`Event`] enum, which encodes the different things that can happen
//! to cause callbacks to run with an `EventContext`.
//!
//! The SpacetimeDB per-module bindings will define a struct `EventContext`,
//! representing a connection to a remote database during some particular event.
//! That `EventContext` struct will have a field `event`, whose type is `Event<Reducer>`,
//! where [`Event`] is defined here and `Reducer` is an enum defined by the per-module bindings.
//!
//! Each callback invoked by the SpacetimeDB SDK will receive an `EventContext` as an argument.
//! You can inspect its `event` field
//! to determine what change in your connection's state caused the callback to run.

use spacetimedb_lib::Timestamp;

use crate::error::InternalError;

#[non_exhaustive]
#[derive(Debug, Clone)]
/// A change in the state of a [`crate::DbContext`] which causes callbacks to run.
pub enum Event<R> {
    /// Event when we are notified that a reducer this client invoked
    /// ran to completion in the remote module and its mutations were committed.
    ///
    /// This event is passed to row callbacks resulting from modifications by the reducer.
    Reducer(ReducerEvent<R>),

    /// Event when one of our subscriptions is applied.
    ///
    /// This event is passed to subscription-applied callbacks,
    /// and to row insert callbacks resulting from the new subscription.
    SubscribeApplied,

    /// Event when one of our subscriptions is removed.
    ///
    /// This event is passed to unsubscribe-applied callbacks,
    /// and to row delete callbacks resulting from the ended subscription.
    UnsubscribeApplied,

    /// Event when a subscription was ended by a disconnection.
    Disconnected,

    /// Event when an error causes one or more of our subscriptions to end prematurely,
    /// or to never be started.
    ///
    /// Payload should be a language-appropriate dynamic error type,
    /// likely `Exception` in C# and `Error` in TypeScript.
    ///
    /// Payload should describe the error in a human-readable format.
    /// No requirement is imposed that it be programmatically inspectable.
    SubscribeError(crate::Error),

    /// Event when we are notified of a transaction in the remote module,
    /// other than one that was the result of a reducer invoked by this client.
    ///
    /// Transactions resulting from reducers invoked by this client will instead report [`Event::Reducer`].
    ///
    /// This event is passed to row callbacks resulting from modifications by the transaction.
    Transaction,
}

#[non_exhaustive]
#[derive(Debug, Clone)]
/// A state change due to a reducer, which may or may not have committed successfully.
pub struct ReducerEvent<R> {
    /// The time at which the reducer was invoked.
    pub timestamp: Timestamp,

    /// Whether the reducer committed, rolled back by returning an error, or was aborted by the host.
    pub status: Status,

    /// The `Reducer` enum defined by the `module_bindings`, which encodes which reducer ran and its arguments.
    pub reducer: R,
}

#[derive(Debug, Clone)]
/// The termination status of a [`ReducerEvent`].
pub enum Status {
    /// The reducer terminated successfully, and its mutations were committed to the database's state.
    Committed,

    /// The reducer returned or threw a handleable error and its mutations were rolled back.
    ///
    /// The `String` payload is the error message signaled by the reducer,
    /// either as an `Err` return, a `panic` message, or a thrown exception.
    Err(String),

    /// The reducer was aborted due to an unexpected or exceptional circumstance.
    Panic(InternalError),
}
